### PR TITLE
Implementation of FLIP-0636 - NFT Metadata

### PR DESCRIPTION
**Pull request by HouseOfHufflepuff**
Implemented FLIP-0636 on the Top Shot contract. Defined a custom view of moment play data called TopShotMomentsMetadataView. Created script, template, and tests to support work. Updated README with implementation details.

### CODE REVIEW

File: lib/go/test/topshot_test.go [hunk 1]
```diff
@@ -398,7 +443,7 @@ func TestMintNFTs(t *testing.T) {
 		assert.Equal(t, cadence.NewBool(true), result)
 
 		result = executeScriptAndCheck(t, b, templates.GenerateGetCollectionIDsScript(env), [][]byte{jsoncdc.MustEncode(cadence.Address(topshotAddr))})
-		idsArray := cadence.NewArray([]cadence.Value{cadence.NewUInt64(6), cadence.NewUInt64(4), cadence.NewUInt64(2), cadence.NewUInt64(1), cadence.NewUInt64(5), cadence.NewUInt64(3)})
+		idsArray := cadence.NewArray([]cadence.Value{cadence.NewUInt64(2), cadence.NewUInt64(3), cadence.NewUInt64(1), cadence.NewUInt64(4), cadence.NewUInt64(5), cadence.NewUInt64(6)})
```
Comment (HouseOfHufflepuff) [inline lib/go/test/topshot_test.go:446]: This is coming back in a different order, now. Adjusted.

File: contracts/TopShot.cdc [hunk 2]
```diff
@@ -455,9 +456,101 @@ pub contract TopShot: NonFungibleToken {
...
+        pub let totalYearsExperience: String?
+        pub let nbaSeason: String?
+        pub let dateOfMoment: String?
+        pub let playCategory: String?
+        pub let playType: String?
+        pub let homeTeamName: String?
+        pub let awayTeamName: String?
+        pub let homeTeamScore: String?
+        pub let awayTeamScore: String?
+        pub let seriesNumber: String?
```
Comment (Deewai) [inline contracts/TopShot.cdc:488]: is there a reason they are all string types? For instance series is a `UInt32` in the contract

File: transactions/scripts/get_nft_metadata.cdc [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+pub fun main(address: Address, id: UInt64): NFT {
+    let account = getAccount(address)
+
+    let collectionRef = account.getCapability(/public/MomentCollection)
+                            .borrow<&{TopShot.MomentCollectionPublic}>()!
+
+    let nft = collectionRef.borrowMoment(id: id)!
+    
+    // Get the basic display information for this NFT
+    let view = nft.resolveView(Type<MetadataViews.Display>())!
```
Comment (Deewai) [inline transactions/scripts/get_nft_metadata.cdc:35]: i thought this script will be returning the `TopShotMomentsMetadataView` view.

**Review by Deewai — COMMENTED** [review]
Great work 👏  
Just some minor questions/comments

File: contracts/TopShot.cdc (hunk 2 above)
Comment (joshuahannan) [reply contracts/TopShot.cdc:488]: Agreed, series number, play ID, set ID, serial number, and the number of moments in this moment's edition should all be included in this, but should be integers and not strings. 👍

**Review by joshuahannan — COMMENTED** [review]
Just one comment on the code about what should be included in the view.

Like @Deewai said, can we also include a script that returns `TopShotMomentsMetadataView` and include some tests for that?

Looks great so far! 👍

File: contracts/TopShot.cdc [hunk 4]
```diff
@@ -455,9 +456,101 @@ pub contract TopShot: NonFungibleToken {
 
     }
 
+    // This is an implementation of a custom metadata view for Top Shot.
+    // This view contains the play metadata.
+    //
+    pub struct TopShotMomentsMetadataView {
```
[the remark this answers is not in the document]
Comment (HouseOfHufflepuff) [reply contracts/TopShot.cdc]: Fixed. Thanks Deewai.

File: contracts/TopShot.cdc (hunk 2 above)
Comment (HouseOfHufflepuff) [reply contracts/TopShot.cdc]: Got it. PR is updated with these changes. Thanks.

File: transactions/scripts/get_nft_metadata.cdc (hunk 3 above)
Comment (HouseOfHufflepuff) [reply transactions/scripts/get_nft_metadata.cdc:35]: Good call. Made a new script for TopShotMomentMetadataView view. Thanks.

File: transactions/scripts/get_topshot_metadata.cdc [hunk 5]
```diff
@@ -0,0 +1,147 @@
+import TopShot from 0xTOPSHOTADDRESS
+import MetadataViews from 0xMETADATAVIEWSADDRESS
+
```
Comment (joshuahannan) [inline transactions/scripts/get_topshot_metadata.cdc:3]: Why define a new struct here when you can just use and return TopShot.TopShotMomentMetadataView?

You could just delete lines 4-100, delete lines 115-146, and just return the view from the contract.

File: transactions/scripts/get_topshot_metadata.cdc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        self.seriesNumber = seriesNumber
+        self.setName = setName
+        self.serialNumber = serialNumber
+        self.playID = playID
+        self.setID = setID
+        self.numMomentsInEdition = numMomentsInEdition
+    }
+}
+
+pub fun main(address: Address, id: UInt64): TopShotMomentMetadataView {
```
Comment (joshuahannan) [inline transactions/scripts/get_topshot_metadata.cdc:102]: return `TopShot.TopShotMomentMetadataView`

File: transactions/scripts/get_topshot_metadata.cdc [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+    let collectionRef = account.getCapability(/public/MomentCollection)
+                            .borrow<&{TopShot.MomentCollectionPublic}>()!
+
+    let nft = collectionRef.borrowMoment(id: id)!
+    
+    // Get the Top Shot specific metadata for this NFT
+    let view = nft.resolveView(Type<TopShot.TopShotMomentMetadataView>())!
+
+    let metadata = view as! TopShot.TopShotMomentMetadataView
```
Comment (joshuahannan) [inline transactions/scripts/get_topshot_metadata.cdc:113]: I don't think the force cast is necessary here

File: transactions/scripts/get_topshot_metadata.cdc [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
```
Comment (joshuahannan) [inline transactions/scripts/get_topshot_metadata.cdc:114]: `return metadata`

**Review by joshuahannan — APPROVED** [review]
Just left some comments about the script. Everything else looks good! 👌

**Review by Deewai — APPROVED** [review]
LGTM!
I agree that we should just return `TopShot.TopShotMomentMetadataView` in the script
Awesome work 🙇

File: transactions/scripts/get_topshot_metadata.cdc (hunk 5 above)
Comment (HouseOfHufflepuff) [reply transactions/scripts/get_topshot_metadata.cdc:3]: Ah, yes. Thank you Josh. This is fixed now.

File: transactions/scripts/get_topshot_metadata.cdc (hunk 8 above)
Comment (HouseOfHufflepuff) [reply transactions/scripts/get_topshot_metadata.cdc:17]: Gotcha. Fixed.

File: transactions/scripts/get_topshot_metadata.cdc (hunk 7 above)
Comment (HouseOfHufflepuff) [reply transactions/scripts/get_topshot_metadata.cdc]: You are right. I have removed. Thanks.

File: transactions/scripts/get_topshot_metadata.cdc (hunk 6 above)
Comment (HouseOfHufflepuff) [reply transactions/scripts/get_topshot_metadata.cdc]: Done. Thank you.

**Pull request merged**